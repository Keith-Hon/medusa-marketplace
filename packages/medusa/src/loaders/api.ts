import { Express } from 'express'
import bodyParser from "body-parser"
import routes from "../api"
import { AwilixContainer } from "awilix"
import { ConfigModule } from "../types/global"

type Options = {
  app: Express
  container: AwilixContainer
  configModule: ConfigModule
}

export default async ({ app, container, configModule }: Options) => {
  app.use(bodyParser.json())
  app.use("/", routes(container, configModule.projectConfig))

  return app
}
