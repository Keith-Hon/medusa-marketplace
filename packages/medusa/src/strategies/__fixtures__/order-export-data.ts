import { DeepPartial } from "typeorm"
import { IdMap } from "medusa-test-utils"
import {
  FulfillmentStatus,
  Order,
  OrderStatus,
  PaymentStatus,
} from "../../models"

const createdAtDate = new Date("2019-01-01T00:00:00.000Z")

export const ordersToExport: DeepPartial<Order>[] = [
  {
    id: "order_1",
    created_at: createdAtDate,
    display_id: 123,
    status: OrderStatus.PENDING,
    fulfillment_status: FulfillmentStatus.PARTIALLY_FULFILLED,
    payment_status: PaymentStatus.CAPTURED,
    subtotal: 10,
    shipping_total: 10,
    discount_total: 0,
    gift_card_total: 0,
    refunded_total: 0,
    tax_total: 5,
    total: 25,
    currency_code: "usd",
    region_id: "region_1",
    shipping_address: {
      id: "address_1",
      address_1: "123 Main St",
      address_2: "",
      city: "New York",
      country_code: "US",
      postal_code: "10001",
    },
    customer: {
      id: "customer_1",
      first_name: "John",
      last_name: "Doe",
      email: "John@Doe.com",
    },
    sales_channel: {
      id: IdMap.getId("sc_1"),
      name: "SC 1",
      description: "SC 1",
    },
  },
  {
    id: "order_2",
    created_at: createdAtDate,
    display_id: 124,
    status: OrderStatus.COMPLETED,
    fulfillment_status: FulfillmentStatus.FULFILLED,
    payment_status: PaymentStatus.CAPTURED,
    subtotal: 125,
    shipping_total: 10,
    discount_total: 0,
    gift_card_total: 0,
    refunded_total: 0,
    tax_total: 0,
    total: 135,
    currency_code: "eur",
    region_id: "region_2",
    shipping_address: {
      id: "address_2",
      address_1: "Hovedgaden 1",
      address_2: "",
      city: "Copenhagen",
      country_code: "DK",
      postal_code: "1150",
    },
    customer: {
      id: "customer_2",
      first_name: "Jane",
      last_name: "Doe",
      email: "Jane@Doe.com",
    },
    sales_channel: {
      id: IdMap.getId("sc_2"),
      name: "SC 2",
      description: "SC 2",
    },
  },
]
