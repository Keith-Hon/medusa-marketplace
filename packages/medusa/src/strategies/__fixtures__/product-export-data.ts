import { IdMap } from "medusa-test-utils"

const productIds = [
  "product-export-strategy-product-1",
  "product-export-strategy-product-2",
]
const variantIds = [
  "product-export-strategy-variant-1",
  "product-export-strategy-variant-2",
  "product-export-strategy-variant-3",
]
export const productsToExport = [
  {
    sales_channels: [
      { id: IdMap.getId("sc_1"), name: "SC 1", description: "SC 1" },
    ],
    collection: {
      created_at: "randomString",
      deleted_at: null,
      handle: "test-collection1",
      id: IdMap.getId("product-export-collection_1"),
      metadata: null,
      title: "Test collection 1",
      updated_at: "randomString",
    },
    collection_id: IdMap.getId("product-export-collection_1"),
    created_at: "randomString",
    deleted_at: null,
    description: "test-product-description-1",
    discountable: true,
    external_id: null,
    handle: "test-product-product-1",
    height: null,
    hs_code: null,
    id: productIds[0],
    images: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-image_1"),
        metadata: null,
        updated_at: "randomString",
        url: "test-image.png",
      },
    ],
    is_giftcard: false,
    length: null,
    material: null,
    metadata: null,
    mid_code: null,
    options: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-option_1"),
        metadata: null,
        product_id: productIds[0],
        title: "test-option-1",
        updated_at: "randomString",
      },
      {
        created_at: "randomString2",
        deleted_at: null,
        id: IdMap.getId("product-export-option_2"),
        metadata: null,
        product_id: productIds[0],
        title: "test-option-2",
        updated_at: "randomString2",
      },
    ],
    origin_country: null,
    profile_id: IdMap.getId("product-export-profile_1"),
    profile: {
      id: IdMap.getId("product-export-profile_1"),
      name: "profile_1",
      type: "profile_type_1",
    },
    status: "draft",
    subtitle: null,
    tags: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-tag_1"),
        metadata: null,
        updated_at: "randomString",
        value: "123_1",
      },
    ],
    thumbnail: null,
    title: "Test product",
    type: {
      created_at: "randomString",
      deleted_at: null,
      id: IdMap.getId("product-export-type_1"),
      metadata: null,
      updated_at: "randomString",
      value: "test-type-1",
    },
    type_id: IdMap.getId("product-export-type_1"),
    updated_at: "randomString",
    variants: [
      {
        allow_backorder: false,
        barcode: "test-barcode",
        calculated_price: null,
        created_at: "randomString",
        deleted_at: null,
        ean: "test-ean",
        height: null,
        hs_code: null,
        id: variantIds[0],
        inventory_quantity: 10,
        length: null,
        manage_inventory: true,
        material: null,
        metadata: null,
        mid_code: null,
        options: [
          {
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-variant_option_1"),
            metadata: null,
            option_id: IdMap.getId("product-export-option_1"),
            updated_at: "randomString",
            value: "option 1 value 1",
            variant_id: variantIds[0],
          },
          {
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-variant_option_2"),
            metadata: null,
            option_id: IdMap.getId("product-export-option_2"),
            updated_at: "randomString",
            value: "option 2 value 1",
            variant_id: variantIds[0],
          },
        ],
        origin_country: null,
        original_price: null,
        prices: [
          {
            amount: 100,
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-price_1"),
            region_id: IdMap.getId("product-export-region_1"),
            max_quantity: null,
            min_quantity: null,
            price_list: null,
            price_list_id: null,
            region: {
              id: IdMap.getId("product-export-region_1"),
              currency_code: "usd",
              name: "france",
            },
            updated_at: "randomString",
            variant_id: variantIds[0],
          },
          {
            amount: 110,
            created_at: "randomString",
            currency_code: "usd",
            deleted_at: null,
            id: IdMap.getId("product-export-price_1"),
            region_id: null,
            max_quantity: null,
            min_quantity: null,
            price_list: null,
            price_list_id: null,
            updated_at: "randomString",
            variant_id: variantIds[0],
          },
          {
            amount: 130,
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-price_1"),
            region_id: IdMap.getId("product-export-region_1"),
            max_quantity: null,
            min_quantity: null,
            price_list: null,
            price_list_id: null,
            region: {
              id: IdMap.getId("product-export-region_3"),
              name: "denmark",
              currency_code: "dkk",
            },
            updated_at: "randomString",
            variant_id: variantIds[0],
          },
        ],
        product_id: IdMap.getId("product-export-product_1"),
        sku: "test-sku",
        title: "Test variant",
        upc: "test-upc",
        updated_at: "randomString",
        weight: null,
        width: null,
      },
    ],
    weight: null,
    width: null,
  },
  {
    sales_channels: [
      { id: IdMap.getId("sc_1"), name: "SC 1", description: "SC 1" },
      { id: IdMap.getId("sc_2"), name: "SC 2", description: "SC 2" },
    ],
    collection: {
      created_at: "randomString",
      deleted_at: null,
      handle: "test-collection2",
      id: IdMap.getId("product-export-collection_2"),
      metadata: null,
      title: "Test collection",
      updated_at: "randomString",
    },
    collection_id: "test-collection",
    created_at: "randomString",
    deleted_at: null,
    description: "test-product-description",
    discountable: true,
    external_id: null,
    handle: "test-product-product-2",
    height: null,
    hs_code: null,
    id: productIds[1],
    images: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-image_2"),
        metadata: null,
        updated_at: "randomString",
        url: "test-image.png",
      },
    ],
    is_giftcard: false,
    length: null,
    material: null,
    metadata: null,
    mid_code: null,
    options: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-option_2"),
        metadata: null,
        product_id: productIds[1],
        title: "test-option",
        updated_at: "randomString",
      },
    ],
    origin_country: null,
    profile_id: IdMap.getId("product-export-profile_2"),
    profile: {
      id: IdMap.getId("product-export-profile_2"),
      name: "profile_2",
      type: "profile_type_2",
    },
    status: "draft",
    subtitle: null,
    tags: [
      {
        created_at: "randomString",
        deleted_at: null,
        id: IdMap.getId("product-export-tag_2"),
        metadata: null,
        updated_at: "randomString",
        value: "123",
      },
    ],
    thumbnail: null,
    title: "Test product",
    type: {
      created_at: "randomString",
      deleted_at: null,
      id: IdMap.getId("product-export-type_2"),
      metadata: null,
      updated_at: "randomString",
      value: "test-type",
    },
    type_id: "test-type",
    updated_at: "randomString",
    variants: [
      {
        allow_backorder: false,
        barcode: "test-barcode",
        calculated_price: null,
        created_at: "randomString",
        deleted_at: null,
        ean: "test-ean",
        height: null,
        hs_code: null,
        id: variantIds[1],
        inventory_quantity: 10,
        length: null,
        manage_inventory: true,
        material: null,
        metadata: null,
        mid_code: null,
        options: [
          {
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-variant_option_2"),
            metadata: null,
            option_id: IdMap.getId("product-export-option_2"),
            updated_at: "randomString",
            value: "Option 1 value 1",
            variant_id: variantIds[1],
          },
        ],
        origin_country: null,
        original_price: null,
        prices: [
          {
            amount: 110,
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-price_2"),
            max_quantity: null,
            min_quantity: null,
            price_list: null,
            price_list_id: null,
            region_id: IdMap.getId("product-export-region_2"),
            region: {
              id: IdMap.getId("product-export-region_2"),
              name: "Denmark",
              currency_code: "dkk",
            },
            updated_at: "randomString",
            variant_id: variantIds[1],
          },
        ],
        product_id: IdMap.getId("product-export-product_2"),
        sku: "test-sku",
        title: "Test variant",
        upc: "test-upc",
        updated_at: "randomString",
        weight: null,
        width: null,
      },
      {
        allow_backorder: false,
        barcode: "test-barcode",
        calculated_price: null,
        created_at: "randomString",
        deleted_at: null,
        ean: "test-ean",
        height: null,
        hs_code: null,
        id: variantIds[2],
        inventory_quantity: 10,
        length: null,
        manage_inventory: true,
        material: null,
        metadata: null,
        mid_code: null,
        options: [
          {
            created_at: "randomString",
            deleted_at: null,
            id: IdMap.getId("product-export-variant_option_2"),
            metadata: null,
            option_id: IdMap.getId("product-export-option_2"),
            updated_at: "randomString",
            value: "Option 1 Value 1",
            variant_id: variantIds[2],
          },
        ],
        origin_country: null,
        original_price: null,
        prices: [
          {
            amount: 120,
            created_at: "randomString",
            currency_code: "usd",
            deleted_at: null,
            id: IdMap.getId("product-export-price_2"),
            max_quantity: null,
            min_quantity: null,
            price_list: null,
            price_list_id: null,
            region_id: IdMap.getId("product-export-region_1"),
            updated_at: "randomString",
            variant_id: variantIds[2],
          },
        ],
        product_id: productIds[1],
        sku: "test-sku",
        title: "Test variant",
        upc: "test-upc",
        updated_at: "randomString",
        weight: null,
        width: null,
      },
    ],
    weight: null,
    width: null,
  },
]
