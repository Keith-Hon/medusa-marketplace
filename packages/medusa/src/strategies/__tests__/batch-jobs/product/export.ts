import ProductExportStrategy from "../../../batch-jobs/product/export"
import { IdMap, MockManager } from "medusa-test-utils"
import { User } from "../../../../models"
import { BatchJobStatus } from "../../../../types/batch-job"
import { productsToExport } from "../../../__fixtures__/product-export-data"
import { AdminPostBatchesReq, defaultAdminProductRelations } from "../../../../api"
import { ProductExportBatchJob } from "../../../batch-jobs/product"
import { Request } from "express"
import { FlagRouter } from "../../../../utils/flag-router";
import SalesChannelFeatureFlag from "../../../../loaders/feature-flags/sales-channels";

const productServiceMock = {
  withTransaction: function () {
    return this
  },
  list: jest.fn().mockImplementation(() => Promise.resolve(productsToExport)),
  count: jest.fn().mockImplementation(() => Promise.resolve(productsToExport.length)),
  listAndCount: jest.fn().mockImplementation(() => {
    return Promise.resolve([productsToExport, productsToExport.length])
  }),
}
const productServiceWithNoDataMock = {
  ...productServiceMock,
  list: jest.fn().mockImplementation(() => Promise.resolve([])),
  count: jest.fn().mockImplementation(() => Promise.resolve(0)),
  listAndCount: jest.fn().mockImplementation(() => {
    return Promise.resolve([[], 0])
  }),
}
const managerMock = MockManager

describe("Product export strategy", () => {
  const outputDataStorage: string[] = []
  const fileServiceMock = {
    delete: jest.fn(),
    getUploadStreamDescriptor: jest.fn().mockImplementation(() => {
      return Promise.resolve({
        writeStream: {
          write: (data: string) => {
            outputDataStorage.push(data)
          },
          end: () => void 0
        },
        promise: Promise.resolve(),
        fileKey: 'product-export.csv'
      })
    }),
    withTransaction: function () {
      return this
    }
  }
  let fakeJob = {
    id: IdMap.getId("product-export-job"),
    type: 'product-export',
    created_by: IdMap.getId("product-export-job-creator"),
    created_by_user: {} as User,
    context: {},
    result: {},
    dry_run: false,
    status: BatchJobStatus.PROCESSING as BatchJobStatus
  } as ProductExportBatchJob

  let canceledFakeJob = {
    ...fakeJob,
    id: "bj_failed",
    status: BatchJobStatus.CANCELED
  } as ProductExportBatchJob

  const batchJobServiceMock = {
    withTransaction: function () {
      return this
    },
    update: jest.fn().mockImplementation((jobOrId, data) => {
      if ((jobOrId?.id ?? jobOrId) === "bj_failed") {
        canceledFakeJob = {
          ...canceledFakeJob,
          ...data,
          context: { ...canceledFakeJob?.context, ...data?.context },
          result: { ...canceledFakeJob?.result, ...data?.result }
        }

        return Promise.resolve(canceledFakeJob)
      }

      fakeJob = {
        ...fakeJob,
        ...data,
        context: { ...fakeJob?.context, ...data?.context },
        result: { ...fakeJob?.result, ...data?.result }
      }

      return Promise.resolve(fakeJob)
    }),
    updateStatus: jest.fn().mockImplementation((status) => {
      fakeJob.status = status
      return Promise.resolve(fakeJob)
    }),
    complete: jest.fn().mockImplementation(() => {
      fakeJob.status = BatchJobStatus.COMPLETED
      return Promise.resolve(fakeJob)
    }),
    retrieve: jest.fn().mockImplementation((id) => {
      const targetFakeJob = id === "bj_failed"
        ? canceledFakeJob
        : fakeJob
      return Promise.resolve(targetFakeJob)
    }),
    setFailed: jest.fn().mockImplementation((...args) => {
      console.error(...args)
    })
  }

  const productExportStrategy = new ProductExportStrategy({
    manager: managerMock,
    fileService: fileServiceMock as any,
    batchJobService: batchJobServiceMock as any,
    productService: productServiceMock as any,
    featureFlagRouter: new FlagRouter({}),
  })

  it('should generate the appropriate template', async () => {
    await productExportStrategy.prepareBatchJobForProcessing(fakeJob, {} as Request)
    await productExportStrategy.preProcessBatchJob(fakeJob.id)
    const template = await productExportStrategy.buildHeader(fakeJob)
    expect(template).toMatch(/.*Product ID.*/)
    expect(template).toMatch(/.*Product Handle.*/)
    expect(template).toMatch(/.*Product Title.*/)
    expect(template).toMatch(/.*Product Subtitle.*/)
    expect(template).toMatch(/.*Product Description.*/)
    expect(template).toMatch(/.*Product Status.*/)
    expect(template).toMatch(/.*Product Thumbnail.*/)
    expect(template).toMatch(/.*Product Weight.*/)
    expect(template).toMatch(/.*Product Length.*/)
    expect(template).toMatch(/.*Product Width.*/)
    expect(template).toMatch(/.*Product Height.*/)
    expect(template).toMatch(/.*Product HS Code.*/)
    expect(template).toMatch(/.*Product Origin Country.*/)
    expect(template).toMatch(/.*Product MID Code.*/)
    expect(template).toMatch(/.*Product Material.*/)
    expect(template).toMatch(/.*Product Collection Title.*/)
    expect(template).toMatch(/.*Product Collection Handle.*/)
    expect(template).toMatch(/.*Product Type.*/)
    expect(template).toMatch(/.*Product Tags.*/)
    expect(template).toMatch(/.*Product Discountable.*/)
    expect(template).toMatch(/.*Product External ID.*/)
    expect(template).toMatch(/.*Product Profile Name.*/)
    expect(template).toMatch(/.*Product Profile Type.*/)
    expect(template).toMatch(/.*Product Profile Type.*/)

    expect(template).toMatch(/.*Variant ID.*/)
    expect(template).toMatch(/.*Variant Title.*/)
    expect(template).toMatch(/.*Variant SKU.*/)
    expect(template).toMatch(/.*Variant Barcode.*/)
    expect(template).toMatch(/.*Variant Allow backorder.*/)
    expect(template).toMatch(/.*Variant Manage inventory.*/)
    expect(template).toMatch(/.*Variant Weight.*/)
    expect(template).toMatch(/.*Variant Length.*/)
    expect(template).toMatch(/.*Variant Width.*/)
    expect(template).toMatch(/.*Variant Height.*/)
    expect(template).toMatch(/.*Variant HS Code.*/)
    expect(template).toMatch(/.*Variant Origin Country.*/)
    expect(template).toMatch(/.*Variant MID Code.*/)
    expect(template).toMatch(/.*Variant Material.*/)

    expect(template).toMatch(/.*Option 1 Name.*/)
    expect(template).toMatch(/.*Option 1 Value.*/)
    expect(template).toMatch(/.*Option 2 Name.*/)
    expect(template).toMatch(/.*Option 2 Value.*/)

    expect(template).not.toMatch(/.*Sales channel 1 Name.*/)
    expect(template).not.toMatch(/.*Sales channel 1 Description.*/)
    expect(template).not.toMatch(/.*Sales channel 2 Name.*/)
    expect(template).not.toMatch(/.*Sales channel 2 Description.*/)

    expect(template).toMatch(/.*Price USD.*/)
    expect(template).toMatch(/.*Price france \[USD\].*/)
    expect(template).toMatch(/.*Price denmark \[DKK\].*/)
    expect(template).toMatch(/.*Price Denmark \[DKK\].*/)

    expect(template).toMatch(/.*Image 1 Url.*/)
  })

  it('should process the batch job and generate the appropriate output', async () => {
    await productExportStrategy.prepareBatchJobForProcessing(fakeJob, {} as Request)
    await productExportStrategy.preProcessBatchJob(fakeJob.id)
    await productExportStrategy.processJob(fakeJob.id)
    expect(outputDataStorage).toMatchSnapshot()
    expect((fakeJob.result as any).file_key).toBeDefined()
  })

  it('should prepare the job to be pre proccessed', async () => {
    const fakeJob1: AdminPostBatchesReq = {
      type: 'product-export',
      context: {
        limit: 10,
        offset: 10,
        expand: "variants",
        fields: "title",
        order: "-title",
        filterable_fields: { title: "test" }
      },
      dry_run: false
    }

    const output1 = await productExportStrategy.prepareBatchJobForProcessing(
      fakeJob1,
      {} as Express.Request
    )

    expect(output1.context).toEqual(expect.objectContaining({
      list_config: {
        select: ["title", "created_at", "id"],
        order: { title: "DESC" },
        relations: ["variants"],
        skip: 10,
        take: 10,
      },
      filterable_fields: { title: "test" }
    }))

    const fakeJob2: AdminPostBatchesReq = {
      type: 'product-export',
      context: {},
      dry_run: false
    }

    const output2 = await productExportStrategy.prepareBatchJobForProcessing(
      fakeJob2,
      {} as Express.Request
    )

    expect(output2.context).toEqual(expect.objectContaining({
      list_config: {
        select: undefined,
        order: { created_at: "DESC" },
        relations: [
          ...defaultAdminProductRelations,
          "variants.prices.region"
        ],
        skip: 0,
        take: 50,
      },
      filterable_fields: undefined
    }))
  })

  it("should always provide a file_key even with no data", async () => {
    const productExportStrategy = new ProductExportStrategy({
      batchJobService: batchJobServiceMock as any,
      fileService: fileServiceMock as any,
      productService: productServiceWithNoDataMock as any,
      manager: MockManager,
      featureFlagRouter: new FlagRouter({}),
    })

    await productExportStrategy.prepareBatchJobForProcessing(fakeJob, {} as Request)
    await productExportStrategy.preProcessBatchJob(fakeJob.id)
    await productExportStrategy.processJob(fakeJob.id)

    expect((fakeJob.result as any).file_key).toBeDefined()
  })

  it("should remove the file_key and file_size if the job is canceled", async () => {
    const productExportStrategy = new ProductExportStrategy({
      batchJobService: batchJobServiceMock as any,
      fileService: fileServiceMock as any,
      productService: productServiceMock as any,
      manager: MockManager,
      featureFlagRouter: new FlagRouter({}),
    })

    await productExportStrategy.prepareBatchJobForProcessing(canceledFakeJob, {} as Request)
    await productExportStrategy.preProcessBatchJob(canceledFakeJob.id)
    await productExportStrategy.processJob(canceledFakeJob.id)

    expect((canceledFakeJob.result as any).file_key).not.toBeDefined()
    expect((canceledFakeJob.result as any).file_size).not.toBeDefined()
  })
})

describe("Product export strategy with sales channels", () => {
  const outputDataStorage: string[] = []
  const fileServiceMock = {
    delete: jest.fn(),
    getUploadStreamDescriptor: jest.fn().mockImplementation(() => {
      return Promise.resolve({
        writeStream: {
          write: (data: string) => {
            outputDataStorage.push(data)
          },
          end: () => void 0
        },
        promise: Promise.resolve(),
        fileKey: 'product-export.csv'
      })
    }),
    withTransaction: function () {
      return this
    }
  }
  let fakeJob = {
    id: IdMap.getId("product-export-job"),
    type: 'product-export',
    created_by: IdMap.getId("product-export-job-creator"),
    created_by_user: {} as User,
    context: {},
    result: {},
    dry_run: false,
    status: BatchJobStatus.PROCESSING as BatchJobStatus
  } as ProductExportBatchJob

  let canceledFakeJob = {
    ...fakeJob,
    id: "bj_failed",
    status: BatchJobStatus.CANCELED
  } as ProductExportBatchJob

  const batchJobServiceMock = {
    withTransaction: function () {
      return this
    },
    update: jest.fn().mockImplementation((jobOrId, data) => {
      if ((jobOrId?.id ?? jobOrId) === "bj_failed") {
        canceledFakeJob = {
          ...canceledFakeJob,
          ...data,
          context: { ...canceledFakeJob?.context, ...data?.context },
          result: { ...canceledFakeJob?.result, ...data?.result }
        }

        return Promise.resolve(canceledFakeJob)
      }

      fakeJob = {
        ...fakeJob,
        ...data,
        context: { ...fakeJob?.context, ...data?.context },
        result: { ...fakeJob?.result, ...data?.result }
      }

      return Promise.resolve(fakeJob)
    }),
    updateStatus: jest.fn().mockImplementation((status) => {
      fakeJob.status = status
      return Promise.resolve(fakeJob)
    }),
    complete: jest.fn().mockImplementation(() => {
      fakeJob.status = BatchJobStatus.COMPLETED
      return Promise.resolve(fakeJob)
    }),
    retrieve: jest.fn().mockImplementation((id) => {
      const targetFakeJob = id === "bj_failed"
        ? canceledFakeJob
        : fakeJob
      return Promise.resolve(targetFakeJob)
    }),
    setFailed: jest.fn().mockImplementation((...args) => {
      console.error(...args)
    })
  }

  const productExportStrategy = new ProductExportStrategy({
    manager: managerMock,
    fileService: fileServiceMock as any,
    batchJobService: batchJobServiceMock as any,
    productService: productServiceMock as any,
    featureFlagRouter: new FlagRouter({
      [SalesChannelFeatureFlag.key]: true,
    }),
  })

  it('should generate the appropriate template', async () => {
    await productExportStrategy.prepareBatchJobForProcessing(fakeJob, {} as Request)
    await productExportStrategy.preProcessBatchJob(fakeJob.id)
    const template = await productExportStrategy.buildHeader(fakeJob)
    expect(template).toMatch(/.*Product ID.*/)
    expect(template).toMatch(/.*Product Handle.*/)
    expect(template).toMatch(/.*Product Title.*/)
    expect(template).toMatch(/.*Product Subtitle.*/)
    expect(template).toMatch(/.*Product Description.*/)
    expect(template).toMatch(/.*Product Status.*/)
    expect(template).toMatch(/.*Product Thumbnail.*/)
    expect(template).toMatch(/.*Product Weight.*/)
    expect(template).toMatch(/.*Product Length.*/)
    expect(template).toMatch(/.*Product Width.*/)
    expect(template).toMatch(/.*Product Height.*/)
    expect(template).toMatch(/.*Product HS Code.*/)
    expect(template).toMatch(/.*Product Origin Country.*/)
    expect(template).toMatch(/.*Product MID Code.*/)
    expect(template).toMatch(/.*Product Material.*/)
    expect(template).toMatch(/.*Product Collection Title.*/)
    expect(template).toMatch(/.*Product Collection Handle.*/)
    expect(template).toMatch(/.*Product Type.*/)
    expect(template).toMatch(/.*Product Tags.*/)
    expect(template).toMatch(/.*Product Discountable.*/)
    expect(template).toMatch(/.*Product External ID.*/)
    expect(template).toMatch(/.*Product Profile Name.*/)
    expect(template).toMatch(/.*Product Profile Type.*/)
    expect(template).toMatch(/.*Product Profile Type.*/)

    expect(template).toMatch(/.*Variant ID.*/)
    expect(template).toMatch(/.*Variant Title.*/)
    expect(template).toMatch(/.*Variant SKU.*/)
    expect(template).toMatch(/.*Variant Barcode.*/)
    expect(template).toMatch(/.*Variant Allow backorder.*/)
    expect(template).toMatch(/.*Variant Manage inventory.*/)
    expect(template).toMatch(/.*Variant Weight.*/)
    expect(template).toMatch(/.*Variant Length.*/)
    expect(template).toMatch(/.*Variant Width.*/)
    expect(template).toMatch(/.*Variant Height.*/)
    expect(template).toMatch(/.*Variant HS Code.*/)
    expect(template).toMatch(/.*Variant Origin Country.*/)
    expect(template).toMatch(/.*Variant MID Code.*/)
    expect(template).toMatch(/.*Variant Material.*/)

    expect(template).toMatch(/.*Option 1 Name.*/)
    expect(template).toMatch(/.*Option 1 Value.*/)
    expect(template).toMatch(/.*Option 2 Name.*/)
    expect(template).toMatch(/.*Option 2 Value.*/)

    expect(template).toMatch(/.*Price USD.*/)
    expect(template).toMatch(/.*Price france \[USD\].*/)
    expect(template).toMatch(/.*Price denmark \[DKK\].*/)
    expect(template).toMatch(/.*Price Denmark \[DKK\].*/)

    expect(template).toMatch(/.*Sales channel 1 Name.*/)
    expect(template).toMatch(/.*Sales channel 1 Description.*/)
    expect(template).toMatch(/.*Sales channel 2 Name.*/)
    expect(template).toMatch(/.*Sales channel 2 Description.*/)

    expect(template).toMatch(/.*Image 1 Url.*/)
  })

  it('should process the batch job and generate the appropriate output', async () => {
    await productExportStrategy.prepareBatchJobForProcessing(fakeJob, {} as Request)
    await productExportStrategy.preProcessBatchJob(fakeJob.id)
    await productExportStrategy.processJob(fakeJob.id)
    expect(outputDataStorage).toMatchSnapshot()
    expect((fakeJob.result as any).file_key).toBeDefined()
  })
})