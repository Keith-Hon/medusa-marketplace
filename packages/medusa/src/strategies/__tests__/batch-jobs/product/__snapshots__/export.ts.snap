// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product export strategy should process the batch job and generate the appropriate output 1`] = `
Array [
  "Product ID;Product Handle;Product Title;Product Subtitle;Product Description;Product Status;Product Thumbnail;Product Weight;Product Length;Product Width;Product Height;Product HS Code;Product Origin Country;Product MID Code;Product Material;Product Collection Title;Product Collection Handle;Product Type;Product Tags;Product Discountable;Product External ID;Product Profile Name;Product Profile Type;Variant ID;Variant Title;Variant SKU;Variant Barcode;Variant Inventory Quantity;Variant Allow backorder;Variant Manage inventory;Variant Weight;Variant Length;Variant Width;Variant Height;Variant HS Code;Variant Origin Country;Variant MID Code;Variant Material;Price france [USD];Price USD;Price denmark [DKK];Price Denmark [DKK];Option 1 Name;Option 1 Value;Option 2 Name;Option 2 Value;Image 1 Url
",
  "product-export-strategy-product-1;test-product-product-1;Test product;;test-product-description-1;draft;;;;;;;;;;Test collection 1;test-collection1;test-type-1;123_1;true;;profile_1;profile_type_1;product-export-strategy-variant-1;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;100;110;130;;test-option-1;option 1 value 1;test-option-2;option 2 value 1;test-image.png
",
  "product-export-strategy-product-2;test-product-product-2;Test product;;test-product-description;draft;;;;;;;;;;Test collection;test-collection2;test-type;123;true;;profile_2;profile_type_2;product-export-strategy-variant-2;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;;;;110;test-option;Option 1 value 1;;;test-image.png
",
  "product-export-strategy-product-2;test-product-product-2;Test product;;test-product-description;draft;;;;;;;;;;Test collection;test-collection2;test-type;123;true;;profile_2;profile_type_2;product-export-strategy-variant-3;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;;120;;;test-option;Option 1 Value 1;;;test-image.png
",
]
`;

exports[`Product export strategy with sales channels should process the batch job and generate the appropriate output 1`] = `
Array [
  "Product ID;Product Handle;Product Title;Product Subtitle;Product Description;Product Status;Product Thumbnail;Product Weight;Product Length;Product Width;Product Height;Product HS Code;Product Origin Country;Product MID Code;Product Material;Product Collection Title;Product Collection Handle;Product Type;Product Tags;Product Discountable;Product External ID;Product Profile Name;Product Profile Type;Variant ID;Variant Title;Variant SKU;Variant Barcode;Variant Inventory Quantity;Variant Allow backorder;Variant Manage inventory;Variant Weight;Variant Length;Variant Width;Variant Height;Variant HS Code;Variant Origin Country;Variant MID Code;Variant Material;Price france [USD];Price USD;Price denmark [DKK];Price Denmark [DKK];Option 1 Name;Option 1 Value;Option 2 Name;Option 2 Value;Image 1 Url;Sales channel 1 Name;Sales channel 1 Description;Sales channel 2 Name;Sales channel 2 Description
",
  "product-export-strategy-product-1;test-product-product-1;Test product;;test-product-description-1;draft;;;;;;;;;;Test collection 1;test-collection1;test-type-1;123_1;true;;profile_1;profile_type_1;product-export-strategy-variant-1;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;100;110;130;;test-option-1;option 1 value 1;test-option-2;option 2 value 1;test-image.png;SC 1;SC 1;;
",
  "product-export-strategy-product-2;test-product-product-2;Test product;;test-product-description;draft;;;;;;;;;;Test collection;test-collection2;test-type;123;true;;profile_2;profile_type_2;product-export-strategy-variant-2;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;;;;110;test-option;Option 1 value 1;;;test-image.png;SC 1;SC 1;SC 2;SC 2
",
  "product-export-strategy-product-2;test-product-product-2;Test product;;test-product-description;draft;;;;;;;;;;Test collection;test-collection2;test-type;123;true;;profile_2;profile_type_2;product-export-strategy-variant-3;Test variant;test-sku;test-barcode;10;false;true;;;;;;;;;;120;;;test-option;Option 1 Value 1;;;test-image.png;SC 1;SC 1;SC 2;SC 2
",
]
`;
